### PR TITLE
BC1.2 charging mode support

### PR DESCRIPTION
**Pull request by keith-zephyr**
Add support for BC1.2 peripherals to act as the charging device. This is a capability often provided on USB Type-C ports.

The BC1.2 charging mode advertised is configured with a devicetree property.

### CODE REVIEW

File: drivers/usb/bc12/bc12_pi3usb9201.c [hunk 1]
```diff
@@ -148,22 +148,49 @@ static void pi3usb9201_notify_callback(const struct device *dev,
...
-	if (state && state->type != BC12_TYPE_NONE) {
-		/* Now update the current charger type */
-		pi3usb9201_data->partner_type = state->type;
+	if (state) {
+		/* Now update callback with the new partner type. */
+		pi3usb9201_data->partner_state = *state;
 		pi3usb9201_notify_callback(dev, state);
 	} else {
-		pi3usb9201_data->partner_type = BC12_TYPE_NONE;
+		pi3usb9201_data->partner_state.bc12_role = BC12_DISCONNECTED;
```
Comment (jfischer-no) [inline drivers/usb/bc12/bc12_pi3usb9201.c:223]: I found the use of the state parameter a bit confusing, NULL is passed only in pi3usb9201_disconnect() to cleanup, right?

File: drivers/usb/bc12/bc12_pi3usb9201.c [hunk 2]
```diff
@@ -110,6 +122,24 @@ static int pi3usb9201_set_mode(const struct device *dev, enum pi3usb9201_mode mo
 				      mode << PI3USB9201_REG_CTRL_1_MODE_SHIFT);
 }
 
+static int pi3usb9201_get_mode(const struct device *dev, enum pi3usb9201_mode *mode)
```
Comment (jfischer-no) [inline drivers/usb/bc12/bc12_pi3usb9201.c:125]:
```suggestion
static int pi3usb9201_get_mode(const struct device *dev, enum pi3usb9201_mode *const mode)
```

**Comment by keith-zephyr** [comment]
Commit https://github.com/zephyrproject-rtos/zephyr/commit/0d0c6d44d3480b8293c0277d36f169b7d4b13af5 is a rebase only.

File: drivers/usb/bc12/bc12_pi3usb9201.c (hunk 1 above)
Comment (keith-zephyr) [reply drivers/usb/bc12/bc12_pi3usb9201.c:230]: Added a description of the function and parameters.

**Review by aaronemassey — DISMISSED** [review]

File: tests/drivers/bc12/src/pd_mode.c [hunk 3]
```diff
@@ -19,15 +19,15 @@
 
 #include <zephyr/logging/log.h>
 
-#define LOG_MODULE_NAME test_bc12
+#define LOG_MODULE_NAME test_bc12_pd
```
Comment (alevkoy) [inline tests/drivers/bc12/src/pd_mode.c:22]: Should this be in the previous commit? Same question for some of the other name changes below.

File: drivers/usb/bc12/bc12_pi3usb9201.c [hunk 4]
```diff
@@ -386,7 +416,7 @@ static int pi3usb9201_init(const struct device *dev)
 	 * Set most recent bc1.2 detection type result to
 	 * BC12_TYPE_NONE for the port.
```
Comment (alevkoy) [inline drivers/usb/bc12/bc12_pi3usb9201.c:576]: Update comment?

**Comment by keith-zephyr** [comment]
Commit https://github.com/zephyrproject-rtos/zephyr/commit/d12b2ff3d9ed366a6d63188aa94c513421083d3c is a rebase only.

File: drivers/usb/bc12/bc12_pi3usb9201.c (hunk 4 above)
Comment (keith-zephyr) [reply drivers/usb/bc12/bc12_pi3usb9201.c]: Done

**Review by alevkoy — APPROVED** [review]

**Review by aaronemassey — APPROVED** [review]

File: drivers/usb/bc12/bc12_pi3usb9201.c [hunk 5]
```diff
@@ -148,22 +148,49 @@ static void pi3usb9201_notify_callback(const struct device *dev,
 	}
 }
 
+static bool pi3usb9201_partner_has_changed(const struct device *dev,
+					   struct bc12_partner_state *state)
```
[the remark this answers is not in the document]
Comment (jfischer-no) [reply drivers/usb/bc12/bc12_pi3usb9201.c]: This is not resolved but is actually very helpful here.

**Review by jfischer-no — APPROVED** [review]

**Review by tmon-nordic — APPROVED** [review]

**Pull request merged**